### PR TITLE
Unseal Wrap trait

### PR DESCRIPTION
**Pull request by hgzimmerman**
Fixes https://github.com/seanmonstar/warp/issues/638.

The doc comments for the `Wrap` trait are not suitable for merging at the moment. Any feedback related to them would be appreciated.

### CODE REVIEW

File: src/filter/wrap.rs [hunk 1]
```diff
@@ -1,27 +1,21 @@
 use super::Filter;
 
-pub trait WrapSealed<F: Filter> {
+/// Wraps filters.
```
Comment (jxs) [inline src/filter/wrap.rs:3]: can we give a detailed explanation of the trait, in the line of others like [reply](https://github.com/seanmonstar/warp/blob/master/src/reply.rs#L198) with a working example?

Comment (hgzimmerman) [reply src/filter/wrap.rs:3]: I'm struggling to come up with an example use case. Most existing uses of the `Wrap` trait use some sealed structure, so I'm not sure what would make sense to demonstrate here.

Comment (jxs) [reply src/filter/wrap.rs]: it's getting nice! can you show where you are stalled? I think you can use `BoxedFilter`

Comment (hgzimmerman) [reply src/filter/wrap.rs]: I'm struggling to come up with a terse, but plausible use for `Wrap`. I look to something like `warp::filters::log::Log` or `warp::filters::reply::WithHeader` and see that their `Wrap` implementation is relying on private items. I don't think `BoxedFilter`s will make this task easier for items that are like `Log` or `WithHeader`

The ultimate motivation behind me wanting to unseal this trait is to be able to implement something like what is detailed in this comment: https://github.com/seanmonstar/warp/pull/408#issuecomment-606325196, but I see that that also relies on other private items.

So its a problem of 1. I don't know what the example should demonstrate. and 2. I don't know if a plausible example could be created with the other "primitives" that are exposed currently.

**Comment by hgzimmerman** [comment]
Ok. So I finally figured out what you meant by using `BoxedFilter`. 
I don't quite have it working yet due to some trait constraints, the errors of which should be present in the CI system, but I'll post the most notable ones here:
```rust
---- src/filter/wrap.rs - filter::wrap::Wrap (line 18) stdout ----
error[E0277]: the trait bound `<<F as warp::filter::FilterBase>::Extract as warp::generic::Tuple>::HList: warp::generic::Combine<_>` is not satisfied
##[error]  --> src/filter/wrap.rs:33:14
   |
18 |             .and(warp::any().map(|| log::info!("")))
   |              ^^^ the trait `warp::generic::Combine<_>` is not implemented for `<<F as warp::filter::FilterBase>::Extract as warp::generic::Tuple>::HList`

error[E0277]: the trait bound `std::convert::Infallible: warp::reject::sealed::CombineRejection<<F as warp::filter::FilterBase>::Error>` is not satisfied
##[error]  --> src/filter/wrap.rs:33:14
   |
18 |             .and(warp::any().map(|| log::info!("")))
   |              ^^^ the trait `warp::reject::sealed::CombineRejection<<F as warp::filter::FilterBase>::Error>` is not implemented for `std::convert::Infallible`
```
Any advice on where to go from here would be appreciated.

**Comment by jxs** [comment]
sorry for the late reply, and thanks for your effort! yeah this solution seems it would always need to unseal other traits which would lead to more confusion, and in the end even if it worked one would always need to impl `Wrap` and probably `Future` which would not be the way one works with `Filter`s.
I have since posted an update on https://github.com/seanmonstar/warp/issues/16#issuecomment-658750842 can you give your opinion there?
Thanks

**Comment by hgzimmerman** [comment]
Should this PR be closed in favor of your `wrap_fn` proposal?

**Comment by hgzimmerman** [comment]
Is there anything I can do to assist in getting your `wrap_fn` included into Warp?

**Comment by jxs** [comment]
@hgzimmerman yeah we can close this one in favor of the `wrap_fn` proposal, which i am awaiting for feedback on the issue.
Thanks

**Pull request closed without merging**